### PR TITLE
metrics: Add metricClusterDeploymentSyncsetPaused

### PR DESCRIPTION
**Pull request by mbarnes**
The OpenShift Dedicated SRE team would like to be alerted when a production ClusterDeployment has a `hive.openshift.io/syncset-pause="true"` annotation for some threshold duration.

This adds a new Prometheus metric to allow for such an alert.  It's essentially a boolean for each ClusterDeployment.

### CODE REVIEW

File: pkg/controller/metrics/metrics.go [hunk 1]
```diff
@@ -192,9 +206,18 @@ func (mc *Calculator) Start(stopCh <-chan struct{}) error {
...
 				}
+
+				var value float64
+				if paused, err := strconv.ParseBool(cd.Annotations[constants.SyncsetPauseAnnotation]); err == nil && paused {
+					value = 1.0
+				}
+				metricClusterDeploymentSyncsetPaused.WithLabelValues(
+					cd.Name,
+					cd.Namespace,
+					clusterType).Set(value)
```
Comment (dgoodwin) [inline pkg/controller/metrics/metrics.go:220]: This is a great idea and thanks for sending us the PR directly!

We want to limit the data we report as much as possible, as implemented it's reporting 0's for every cluster in the system, I think you can just skip the 0's.

Then you would want to make sure to delete any metric that shouldn't be set (for clusters that were paused but no longer are) so it stops getting reported (and eventually cycles out of the prometheus data as it expires a week or two later), which you can do idempotently with this approach:
https://github.com/openshift/hive/blob/master/pkg/controller/clusterdeployment/clusterdeployment_controller.go#L1927

Comment (mbarnes) [reply pkg/controller/metrics/metrics.go:220]: That makes sense, and thanks for the quick review.  I'll rework as you suggest.

**Comment by mbarnes** [comment]
@dgoodwin Reworked as you suggested, though CI tests seem to be wedged at the moment.

This version is a little more intrusive as I had to split a util function into two and tend to its unit tests.

File: pkg/controller/metrics/metrics.go [hunk 2]
```diff
@@ -146,6 +159,19 @@ func Add(mgr manager.Manager) error {
...
+// ClearClusterDeploymentSyncsetPaused deletes the MetricClusterDeploymentSyncsetPaused
+// labeled value for the given ClusterDeployment.
+func ClearClusterDeploymentSyncsetPaused(cd *hivev1.ClusterDeployment, logger log.FieldLogger) {
+	cleared := MetricClusterDeploymentSyncsetPaused.Delete(map[string]string{
+		"cluster_deployment": cd.Name,
+		"namespace":          cd.Namespace,
+		"cluster_type":       GetClusterDeploymentType(cd),
+	})
+	if cleared {
+		logger.Debug("cleared metric: %v", MetricClusterDeploymentSyncsetPaused)
```
Comment (dgoodwin) [inline pkg/controller/metrics/metrics.go:171]: Probably worth Info level as this should be pretty rare.

File: pkg/controller/metrics/metrics.go [hunk 3]
```diff
@@ -192,9 +219,16 @@ func (mc *Calculator) Start(stopCh <-chan struct{}) error {
...
+						clusterType).Set(
 						time.Since(cd.CreationTimestamp.Time).Seconds())
 				}
+
+				if paused, err := strconv.ParseBool(cd.Annotations[constants.SyncsetPauseAnnotation]); err == nil && paused {
+					MetricClusterDeploymentSyncsetPaused.WithLabelValues(
+						cd.Name,
+						cd.Namespace,
+						clusterType).Set(1.0)
+				}
```
Comment (dgoodwin) [inline pkg/controller/metrics/metrics.go:231]: I might be mistaken, but I think you can just do the Delete in an else right here, does that work? I fear my link misled you to plug it into a controller, whereas in this case I think it's safe for us to clear it right here. (was intended to demonstrate how to Delete for specific labels)

File: pkg/controller/utils/clusterdeployment.go [hunk 4]
```diff
@@ -23,17 +23,21 @@ func IsFakeCluster(cd *hivev1.ClusterDeployment) bool {
...
+func IsClusterPaused(cd *hivev1.ClusterDeployment, logger log.FieldLogger) bool {
 	if paused, err := strconv.ParseBool(cd.Annotations[constants.SyncsetPauseAnnotation]); err == nil && paused {
 		logger.WithField("annotation", constants.SyncsetPauseAnnotation).Warn("syncing to cluster is disabled by annotation")
 		return true
 	}
+	return false
+}
+
+// IsClusterRelocating checks if the cluster is relocating
+func IsClusterRelocating(cd *hivev1.ClusterDeployment, logger log.FieldLogger) bool {
```
Comment (dgoodwin) [inline pkg/controller/utils/clusterdeployment.go:36]: If the above is correct, the rest of this refactor can be reverted back to original form.

**Comment by dgoodwin** [comment]
:100: 

Thanks!

/lgtm

**Pull request merged**